scripts/spack/packages/serac/package.py

```diff
@@ -178,17 +181,24 @@ class Serac(CachedCMakePackage, CudaPackage):
     depends_on("sundials build_type=Debug".format(dep), when="+sundials build_type=Debug".format(dep))
 
     # Optional (require when="+profile")
-    for dep in ["adiak", "caliper"]:
-        depends_on("{0} build_type=Debug".format(dep), when="+profiling build_type=Debug")
-        depends_on("{0}+shared".format(dep), when="+profiling+shared")
-        depends_on("{0}~shared".format(dep), when="+profiling~shared")
+    depends_on("adiak build_type=Debug".format(dep), when="+profiling build_type=Debug")
+    depends_on("adiak+shared".format(dep), when="+profiling+shared")
+    depends_on("adiak~shared".format(dep), when="+profiling~shared")
+
+    # Don't propagate ~shared to caliper in rocm builds
+    depends_on("caliper build_type=Debug".format(dep), when="+profiling build_type=Debug")
+    depends_on("caliper+shared".format(dep), when="+profiling+shared")
+    depends_on("caliper~shared".format(dep), when="+profiling~shared~rocm")
```

Comment:
Should this be two lines?
```suggestion
    depends_on("caliper~shared".format(dep), when="+profiling~shared")
    depends_on("caliper~shared".format(dep), when="~rocm")
```

Reply:
i tried that and unfortunately it doesn't concretize. i think it all needs to be on one line to work around static caliper not building on rocm and also allowing forwarding `shared` variant to caliper on other.

```
==> Error: failed to concretize `serac@develop%clang@17.0.0~openmp~petsc+profiling+raja+rocm+strumpack+umpire amdgpu_target=gfx90a ^hdf5 cflags=-Wno-int-conversion` for the following reasons:
     1. cannot satisfy a requirement for package 'caliper'.
```

Reply:
btw, this is the caliper build error i get with `caliper~shared`

```
  >> 605    ld.lld: error: undefined symbol: hipKernelNameRef
     606    >>> referenced by RocTracer.cpp
     607    >>>               RocTracer.cpp.o:((anonymous namespace)::RocTracerService::hip_api_callback(unsigned int, unsigned int, void const*, void*)) in archive ../../libcaliper-serial.
            a
     608    
  >> 609    ld.lld: error: undefined symbol: hipKernelNameRefByPtr
     610    >>> referenced by RocTracer.cpp
     611    >>>               RocTracer.cpp.o:((anonymous namespace)::RocTracerService::hip_api_callback(unsigned int, unsigned int, void const*, void*)) in archive ../../libcaliper-serial.
            a
  >> 612    clang-16: error: linker command failed with exit code 1 (use -v to see invocation)
```